pkg/pub_package_reader/lib/pub_package_reader.dart

```diff
@@ -466,6 +468,12 @@ Iterable<ArchiveIssue> checkSdkVersionRange(Pubspec pubspec) sync* {
     yield ArchiveIssue(
         'The SDK constraint allows Dart 4.0.0, this is not allowed because Dart 4 does not exist.');
   }
+
+  if (!sdk.intersect(_preNullSafety).isEmpty &&
+      !sdk.intersect(_postDart3).isEmpty) {
+    yield ArchiveIssue(
+        'The SDK lower constraint is outside of the Dart 3 compatibility window (">=2.12.0").');
```

Comment:
```suggestion
        'An SDK lower bound constraint less than 2.12.0 is not compatible with Dart 3.\n'
        'Either bump the lower bound SDK constraint to `>=2.12.0`, or\n'
        'lower the upper bound SDK constraint to `<3.0.0`.');
```

Reply:
Or something like that... we could possibly create the constraints and write them out to make it more obvious.